20221103-endpoint-authz.md

```diff
@@ -0,0 +1,304 @@
+
+## Configuration validation
+The only validation to perform is on the keys indicating protected endpoints. When parsing the configuration, we can compare the provided keys to the RPC methods defined in the [OpenFGAService](https://github.com/openfga/api/blob/main/openfga/v1/openfga_service.proto#L14). If a key does not exist in the service, then the configuration will be rejected.
+
+## Authorization evaluation
+The [`Authenticator` interface](https://github.com/openfga/openfga/blob/main/server/authn/authn.go#L21-L28) will be extended, adding an `Authorize` function:
+
+```go
+type Authenticator interface {
+	// Authenticate returns a nil error and the AuthClaims info (if available) if the subject is authenticated or a
+	// non-nil error with an appropriate error cause otherwise.
+	Authenticate(requestContext context.Context) (*AuthClaims, error)
+
+	// Authorize returns a nil error if the subject is authorized or a non-nil error with an appropriate error cause
+	// otherwise. It requires that the context has been augmented with AuthClaims
+	Authorize(requestContext context.Context, fullPath string) error
+
+	// Close Cleans up the authenticator.
+	Close()
+}
+```
+
+As the comment indicates, the `Authorize` function will rely on the `Authenticate` function augmenting the request context with the claims extracted from the `Authorization` header. Additionally, it requires the full method path to be supplied in order to compare against the configured protected endpoints.
+
+The [`AuthFunc` middleware](https://github.com/openfga/openfga/blob/main/internal/middleware/authn.go#L10-L19) will be augmented to support performing an authorization check. In order to access the full request method, it must be constructed the `grpc.UnaryServerInterceptor` directly instead of via `github.com/grpc-ecosystem/go-grpc-middleware/auth`. The middleware will first perform the authentication check with `Authenticate` and augment the request context. If there are no errors, it will then invoke `Authorize`.
+
+Inside of `Authorize`, evaluation is performed as follows:
+1. Extract the method name from the fullPath and lowercase.
+2. If the method name exists in the `ProtectedEndpoints` config:
+   1. If either the scope or the subject exist in the protected endpoint config for the method name, then return `nil`.
+3. If there is a global config:
+   1. If either the scope or the subject exist in the global config, then return `nil`.
+4. Return an authorization error.
+
+# Migration
+[migration]: #migration
+
+No migration is needed; support for endpoint authorization will be completely opt-in.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* If the implementation is simple, it will not support all authorization use-cases (i.e. combinations of allow and deny).
+* If the implementation allows for more complex authorization schemes, it will be more difficult to configure and more prone to implementation bugs.
+
+# Alternatives
```

Comment:
Would it make sense to consider not to include `subjects`? 

Given you'll need a OIDC server, you can take advantage of the ability of the server to assign scopes per client_id. 

I'm not familiar with how this works with other OIDC servers but I know Auth0 supports it. Do you know if other OIDC servers do not have this feature?

Reply:
Yeah, that would be reasonable! I included the subjects as an option as the code currently extracts both the subject and the scopes. I think that most operators would likely authorize based on scopes, I think we could leave the `subjects` portion out unless someone asks for it.